### PR TITLE
User/lemccomb/fileread

### PR DESCRIPTION
**Pull request by lemccomb**
1. Makes CoseSignTool and CoseHandler more resilient to sign and validate requests on files that may not have been fully written yet or fully disposed.
2. Forces full dispose of streams at the end of sign and validate operations.
3. Implements a simple logger to write to console, trace, or other target.

### CODE REVIEW

File: CoseHandler/Extensions/FileInfoExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    /// <returns>The file content.</returns>
+    public static FileStream? GetStreamResilient(this FileInfo f, OutputTarget? writeTo = null) => GetBytesOrStream(f, true, writeTo).Item2;
+
+    private static (byte[]?, FileStream?) GetBytesOrStream(FileInfo f, bool isStream, OutputTarget? writer)
+    {
+        // Make sure the file exists, allowing retries in case it hasn't hit the disk yet.
+        DateTime startTime = DateTime.Now;
+        while (f is null || !f.Exists)
+        {
+            Thread.Sleep(100);
```
Comment (elantiguamsft) [inline CoseHandler/Extensions/FileInfoExtensions.cs:35]: should we define this backoff time value in a constant somewhere? I noticed we are using 250 below as well, so might be useful to have a comment or variable to describe why these values are what they are. Might also make sense to allow the user to control this value (within an acceptable range), with the current values as defaults - same for the max wait time passed into `OutOfTime()`

File: CoseHandler/Extensions/FileInfoExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            Thread.Sleep(250);
+            if (OutOfTime(startTime, 5) && lastLength == f.Length)
+            {
+                if (f.Length > lastLength)
+                {
+                    ticks++;
+                    if (!waitMessageStarted)
+                    {
+                        waitMessageStarted = true;
+                        writer.WriteLine($"Waiting for write of file '{f.FullName}' to complete.");
```
Comment (elantiguamsft) [inline CoseHandler/Extensions/FileInfoExtensions.cs:73]: love the added output information! Should we put it behind the verbose flag?

File: CoseSignTool.Tests/ValidateCommandTests.cs [hunk 3]
```diff
@@ -155,39 +155,73 @@ public void ValidateUntrustedFails()
...
         coseFile.Seek(0, SeekOrigin.Begin);
 
         // setup validator
         var validator = new ValidateCommand();
-        var result = validator.RunCoseHandlerCommand(coseFile,
-                                                     new FileInfo(payloadFile),
-                                                     [SelfSignedCert],
-                                                     X509RevocationMode.Online,
-                                                     null,
-                                                     false);
```
Comment (elantiguamsft) [inline CoseSignTool.Tests/ValidateCommandTests.cs:178]: I may be confused, but I thought these were required params for `RunCoseHandlerCommand()`?

Comment (lemccomb) [reply CoseSignTool.Tests/ValidateCommandTests.cs:178]: Nope. They have default values.

File: CoseHandler/Extensions/FileInfoExtensions.cs (hunk 2 above)
Comment (lemccomb) [reply CoseHandler/Extensions/FileInfoExtensions.cs:76]: I don't think so. If I was having to wait on files not yet fully written I'd want to know about it.

File: CoseHandler/Extensions/FileInfoExtensions.cs (hunk 1 above)
Comment (lemccomb) [reply CoseHandler/Extensions/FileInfoExtensions.cs:39]: Done.

**Review by elantiguamsft — APPROVED** [review]

**Pull request merged**